src/controller/OperationalCredentialsDelegate.h

```diff
@@ -0,0 +1,99 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/util/basic-types.h>
+#include <crypto/CHIPCryptoPAL.h>
+#include <support/DLLUtil.h>
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+namespace Controller {
+
+/// Callbacks for CHIP operational credentials generation
+class DLL_EXPORT OperationalCredentialsDelegate
+{
+public:
+    virtual ~OperationalCredentialsDelegate() {}
+
+    /**
+     * @brief
+     *   This function generates an operational certificate for the given node.
+     *   The API generates the certificate in X.509 DER format.
+     *
+     *   The delegate is expected to use the certificate authority whose certificate
+     *   is returned in `GetIntermediateCACertificate()` or `GetRootCACertificate()`
+     *   API calls.
+     *
+     * @param[in] nodeId       Node ID of the target device.
+     * @param[in] fabricId     Fabric ID for which the credentials are being generated.
+     * @param[in] csr          Certificate Signing Request from the node in DER format.
+     * @param[in] csrLength    The length of Certificate Signing Request.
+     * @param[in] serialNumber Serial number to assign to the new certificate.
+     * @param[in] certBuf      The API will fill in the generated cert in this buffer. The buffer is allocated by the caller.
+     * @param[in] certBufSize  The size of certBuf buffer.
+     * @param[out] outCertLen  The size of the actual certificate that was written in the certBuf.
+     *
+     * @return CHIP_ERROR CHIP_NO_ERROR on success, or corresponding error code.
+     */
+    virtual CHIP_ERROR GenerateNodeOperationalCertificate(NodeId nodeId, FabricId fabricId, const uint8_t * csr, size_t csrLength,
```

Comment:
8 arguments is a lot. Consider grouping:

- PeerId instead of Nodeid + fabricId
- ByteSpan for csr and certbuf

This would make at least this method have 5 parameters instead of 8 and would group related items.

Reply:
The latest patch changes CSR to bytes pan. I'll update the NodeID/FabricID as well.